matching-service/controller/matching-controller.js

```diff
@@ -0,0 +1,258 @@
+
+                while (queueSize > 0 && cancelMatchmakeUsers.has(firstOpponent[0]) && cancelMatchmakeUsers.get(firstOpponent[0]) > firstOpponent[2])
+                {
+                    firstOpponent = await redis.lpop(queueName);
+                    firstOpponent = firstOpponent.split(":");
+                    queueSize--;
+                }
+
+
+                if (queueSize == 0) 
+                {
+                    // if not ttl, put the user back in queue
+                    if (!cancelMatchmakeUsers.has(firstOpponent[0]) || cancelMatchmakeUsers.get(firstOpponent[0]) < firstOpponent[2]) 
+                    {
+                        redis.rpush(queueName, firstOpponent.join(":"));
+                        queueSize++;
+                    }
+
+                    break;
+                }
+                let secondOpponent = await redis.lpop(queueName);
+                secondOpponent = secondOpponent.split(":");
+                queueSize--;
+
+                while (queueSize > 0 && cancelMatchmakeUsers.has(secondOpponent[0]) && cancelMatchmakeUsers.get(secondOpponent[0]) > secondOpponent[2])
+                {
+                    secondOpponent = await redis.lpop(queueName);
+                    secondOpponent = secondOpponent.split(":");
+                    queueSize--;
+                }
+
+                if (cancelMatchmakeUsers.has(secondOpponent[0]) && cancelMatchmakeUsers.get(secondOpponent[0]) > secondOpponent[2])
+                {
+                    redis.rpush(queueName, firstOpponent.join(":"));
+                    queueSize++;
+                    break;
+                }
+
+                emitMatchFound(firstOpponent, secondOpponent, category, difficulty);
+            }
+
+            if (queueSize == 1)
+            {
+                let remainingUser = await redis.lpop(queueName)
+                remainingUser = remainingUser.split(":");
+
+                // * check if there's any player from loosen_difficulty queue to match with 
+                const noDifficultyQueueSize = await redis.llen(category);
+
+                if (noDifficultyQueueSize > 0) 
+                {
+                    let opponent = await redis.lpop(category);
+                    opponent = opponent.split(":");
+                    emitMatchFound(opponent, remainingUser, category, difficulty);
+                    continue;
+                }
+
+                const time = Date.now() - parseInt(remainingUser[2]);
+
+                if (time > LOOSEN_DIFFICULTY_TIME) 
+                {   
+                    redis.rpush(category,`${remainingUser.join(":")}:${difficulty}`);
+                    io.to(remainingUser[1]).emit("loosen-difficulty", remainingUser[0]);
+                }
+                else 
+                {
+                    redis.rpush(queueName, remainingUser.join(":"));
+                }
+            }
+            
+        }
+
+    }
+    
+}
+
+setInterval(matchUserInQueue, CHECK_INTERVAL);
```

Comment:
Can we put this somewhere else? Don't think its good to call a method in the controller.

Reply:
Thanks for the suggestion!